crates/ruff_cli/src/commands/run.rs

```diff
@@ -144,6 +147,44 @@ pub fn run(
             acc
         });
 
+    debug!("{:#?}", diagnostics.imports);
+    let mut cycles: FxHashMap<Arc<str>, FxHashSet<Vec<Arc<str>>>> = FxHashMap::default();
+
+    for (path, package, settings) in
+        paths
+            .iter()
+            .filter_map(|entry| entry.as_ref().ok())
+            .map(|entry| {
+                let path = entry.path();
+                let package = path
+                    .parent()
+                    .and_then(|parent| package_roots.get(parent))
+                    .and_then(|package| *package);
+                let settings = resolver.resolve_all(path, pyproject_strategy);
+                (path, package, settings)
+            })
+    {
+        if settings.lib.rules.enabled(Rule::CyclicImport) {
+            if let Some(cycle_diagnostics) =
+                pylint_cyclic_import(path, package, &diagnostics.imports, &mut cycles)
+            {
+                diagnostics += Diagnostics::new(
+                    cycle_diagnostics
+                        .into_iter()
+                        .map(|diagnostic| {
+                            Message::from_diagnostic(
+                                diagnostic,
+                                SourceFileBuilder::new(&path.to_string_lossy()).finish(),
```

Comment:
Can you test if ruff correctly renders the diagnostic using `--show-source`. You may need to set the source code if this is not the case.

Reply:
The answer is that it doesn't render it correctly so I'll need to sort this out.

Reply:
I've got this working, but I'm not sure I've done this in the expected fashion. If I want to use `ruff::linter::diagnostics_to_messages` I have to re-tokenize the source file which seems inefficient. I would need to do that to populate the `noqa` flags correctly, should I do that?

Reply:
I plan to have a closer look and provide you with an answer by Friday or early next week (I'm travelling today/tomorrow)